### DISCUSSION TITLE
Inquiry Regarding Availability of Position-Based Fluid and Rigid Body Solvers in NVIDIA PhysX Source Code

### DISCUSSION THREAD

**Discussion by 123uiui** [NVIDIA-Omniverse/PhysX#327] [Q&A]
Dear NVIDIA PhysX Development Team,
I hope this message finds you well. I am currently exploring the possibility of utilizing NVIDIA's PhysX framework for some vehicle in water scene. However, upon reviewing the available source code on GitHub, I was unable to find specific modules related to the construction and implementation of position based fluid models （or the so called particle based dynamics PBD model), and the rigid body constraint solver.
Could you please clarify whether the current PhysX source code includes the necessary components and algorithms for the simulation of position based fluid and rigid body collision solver (PGS or TGS solver)? Specifically, I would like to understand if the source code contains any foundational model for solving fluid and rigid body governing equations or if the publicly available GitHub repository only includes the SDK without the detailed source code for fluid and rigid body simulation.
Your reply would be greatly appreciated. Thank you very much for your time and assistance.
Best regards

**Comment by vreutskyy** [comment]
Hi @123uiui 
Thank you for your inquiry about PhysX. The current PhysX source code on GitHub primarily contains the CPU implementation of the framework. While it includes robust support for rigid body dynamics, including collision detection and resolution, the fluid simulation features you're interested in are actually GPU-only and not part of the open-source CPU code.
The position-based fluid models and advanced particle-based dynamics (PBD) are incorporated from the former NVIDIA Flex library, but these are only available through pre-compiled GPU binaries and not in the open-source repository. Similarly, while the rigid body constraint solver is included in the CPU source, the GPU-accelerated version is part of the closed-source GPU code.
Regards
